### DISCUSSION TITLE
Basic Help.

### DISCUSSION THREAD

**Discussion by DFergFLA** [XusinboyBekchanov/VisualFBEditor#166] [General]
I am so confused.  I have tried a few things with this.  I can change the label on a button.
CommandButton1.text = "Hello" works
But I can't change the label of a form
From1.text = "hello" 
doesn't work
Form1.capttion = "hello"
doesn't work.

I get compile errors like "error 214: Member isn't static" or "error255 Ambiguous symbol access.  Explicate scope resolution required,  found "text"

Neither one of those mean anything to me and I haven't been able to find any information using google.  Is there any documentation at all for these forms, widgets?  I'm not sure I understand what the difference between, Text, and Caption.   But it works for the command button but not a form.  It would seem to me that it should work the same way as the text for the button.

Frustration aside.  I have been able to do some simple loops and math with output to the terminal.  Overall I am really having fun.  But being a newbie my ability to understand the problems and troubleshoot them is very limited.

**Comment by XusinboyBekchanov** [comment]
Form1 is a type. You must refer to fForm1, below is the definition:
`Dim Shared fForm1 As Form1`

**Reply by W98X** [reply]
Why can't it be called Tform1 like this:

```
Type TForm1 Extends Form
	Declare Constructor
		
	End Type

Dim Shared Form1 As TForm1

form1.show
form1.caption = "I am Form 1"
```

By using `Tform1` for the type means that `form1` is available to use as `form1.show` and is more intuitive, obvious and newbie freindly and **doesn't need any documentation or explaination**. 

Whereas `fForm1` is not intuitive, not obvious and newbies don't understand why an extra `f` has to be put infront of `form1`.

When I tried using `form1.show` for the first time, I hit the same error as DFergFLA, and I'm sure every single newbie will hit that same error the first time. Because `form1` is obvious and `fForm1` isn't.

**Reply by demosthenesk** [reply]
https://github.com/XusinboyBekchanov/VisualFBEditor/discussions/166#discussioncomment-1297753

**Reply by demosthenesk** [reply]
A form may could be named as fSave which means that there is a Save dialog form, for example.
fSave means that we have a form (f) with name Save = fSave.
It's the same like an integer variable with name iVar, or a double variable with name dVar, or a CommandButton with name btnLoad.
This kind of naming variables is common and it is used by many developers.

**Reply by XusinboyBekchanov** [reply]
I, too, now analyze all the pros and cons, then I will come up with some kind of solution.

**Reply by XusinboyBekchanov** [reply]
Improved: Changed Form Type name and Form name:
https://github.com/XusinboyBekchanov/VisualFBEditor/commit/bcccbd0fa274b8504e1cc9b21fa40c53b5dd4d48

**Reply by demosthenesk** [reply]
> Improved: Changed Form Type name and Form name:
> [bcccbd0](https://github.com/XusinboyBekchanov/VisualFBEditor/commit/bcccbd0fa274b8504e1cc9b21fa40c53b5dd4d48)

Is it possible to have a choice between old template and new one?
I liked more the old template and my examples in the book are based on the old template.
Thanks.

**Reply by XusinboyBekchanov** [reply]
Works in the old template too. In my opinion, only we need to make options in the Designer settings or elsewhere.

**Reply by demosthenesk** [reply]
> Works in the old template too. In my opinion, only we need to make options in the Designer settings or elsewhere.

Yes, that i said to have an option from settings.

**Reply by XusinboyBekchanov** [reply]
If you copy the old templates to folder it works.

**Reply by demosthenesk** [reply]
> If you copy the old templates to folder it works.

it will overwrite the existing Form.frm which is the new one template.
The best is to have two Form templates and an option in settings, so the functionality exists from downloaded master-zip file.

**Reply by XusinboyBekchanov** [reply]
Added: "Create Form types without Type word" option to Designer options:
https://github.com/XusinboyBekchanov/VisualFBEditor/commit/1cb24b4e8eb32889e3ad265ad8cf4102652935b9

**Reply by demosthenesk** [reply]
> Added: "Create Form types without Type word" option to Designer options:
> [1cb24b4](https://github.com/XusinboyBekchanov/VisualFBEditor/commit/1cb24b4e8eb32889e3ad265ad8cf4102652935b9)

ok!
Nice, it works!
Thanks!

**Reply by XusinboyBekchanov** [reply]
Thank you too.

**Reply by W98X** [reply]
> 
> 
> Improved: Changed Form Type name and Form name:
> [bcccbd0](https://github.com/XusinboyBekchanov/VisualFBEditor/commit/bcccbd0fa274b8504e1cc9b21fa40c53b5dd4d48)

That's great! Now the form can be accessed according to the name the user chooses, with no need for an unexplained `f` at the start. 👍 

Now the user is able to chose any naming convention they want (or no naming convention at all!). This is achieved via the form name property in the property grid.

In the screenshot below, I have used the naming convention `frm` at the start of the form name and now it can be accessed as `frmMyForm1.show`

![vfbr form name convention frm](https://user-images.githubusercontent.com/88371717/132680400-a21001f1-b05d-4b1b-ada7-70af6c6e2d0e.png)

In another screenshot, I have used the naming convention of `f` at the start, to make it accessible via `fMyForm1.show`

![vfbe form name with f](https://user-images.githubusercontent.com/88371717/132680525-3e4d14a0-05fd-444d-82d9-d588d36688b2.png)

In another screenshot, I have used no naming convention at all. The form is accessible according the name I gave it in the property panel `MyForm1`. So now it is easy and obvious to access the form via `MyForm1.show` or `MyForm1.caption = "I am form 1"`

![vfbe form no naming conventon](https://user-images.githubusercontent.com/88371717/132681361-9dac7164-e336-4efd-9809-1456247ed5ba.png)
 
From this day onwards, newbies will no longer hit the error that DFergFLA experienced of a form name needing an unexplained `f` at the start. The form will now be accessible according to the name it has in the properties panel. This is obvious and easy now.

This is a big improvement. Many thanks. 😀

**Reply by XusinboyBekchanov** [reply]
Good.

**Reply by XusinboyBekchanov** [reply]
Fixed: Name property of Form in Designer:
https://github.com/XusinboyBekchanov/VisualFBEditor/commit/73225be923d6f41f6941cbc866912e948adcc377

**Comment by DFergFLA** [comment]
Ok, , where would that go?

**Comment by XusinboyBekchanov** [comment]
By default, this is already there:
```
#ifdef __FB_WIN32__
	'#Compile "Form1.rc"
#endif
'#Region "Form"
	#include once "mff/Form.bi"
	
	Using My.Sys.Forms
	
	Type Form1 Extends Form
		
	End Type
	
	Dim Shared fForm1 As Form1
	
	#ifndef _NOT_AUTORUN_FORMS_
		fForm1.Show
		
		App.Run
	#endif
'#End Region
```

**Comment by DFergFLA** [comment]
Ok, I can see that code in the form1 constructor.   So, I am still at square 1.  If that is the code I need to make this work and that code is already there.... Then why can't I change the form1.text ?

**Comment by DFergFLA** [comment]
![Screenshot 2021-09-07 213611](https://user-images.githubusercontent.com/32622501/132431857-b7337525-894d-41de-a47a-dc0ffbb4bf0c.png)

**Comment by XusinboyBekchanov** [comment]
Here:
![image](https://user-images.githubusercontent.com/32607344/132433160-c92500ef-3fe1-4b62-8911-d3710457cec7.png)

**Comment by DFergFLA** [comment]
Thank You, thank you, thank you.

Can I get a little more info on what the lower case f in "fForm1.text" is for?

**Comment by XusinboyBekchanov** [comment]
Variables can be named in different ways:
![image](https://user-images.githubusercontent.com/32607344/132434945-d5a1df12-7426-4c37-a335-1d2a56b9c30d.png)

**Reply by demosthenesk** [reply]
Every developer or company has their own standards for code writing.
For example for standard data types like integer, long, double etc i name variables like iVar, lVar, dVar
For custom data types like forms, buttons, comboboxes, lists etc i name variable objects like fForm1, btnSave, cmbList, lstList etc...
You can define your own name standards as you like !

**Comment by demosthenesk** [comment]
### Anatomy of a Form1.frm code
Every form has a template code for the beginning when you add a form in your project.
The following code is an example:

```
#ifdef __FB_WIN32__
	'#Compile "Form1.rc"
#endif
'#Region "Form"
	#include once "mff/Form.bi"
	
	Using My.Sys.Forms
	
	Type Form1 Extends Form
		Declare Constructor
		
	End Type
	
	Constructor Form1
		' Form1
		With This
			.Name = "Form1"
			.Text = "Form1"
			.Caption = "Form1"
			.SetBounds 0, 0, 350, 300
		End With
	End Constructor
	
	Dim Shared fForm1 As Form1
	
	#ifndef _NOT_AUTORUN_FORMS_
		fForm1.Show
		
		App.Run
	#endif
'#End Region
```
First we have a preprocessor check to test if the code runs in windows. If it is, then we order to compile also Form1.rc file.

```
#ifdef __FB_WIN32__
	'#Compile "Form1.rc"
#endif

```
After that we have the Region form code.
We include header files from mff according to what kind of objects we have on the Form.
Below we include only the form header.
And we use the Forms namespace.

```
	#include once "mff/Form.bi"
	
	Using My.Sys.Forms

``` 
After we have the Form1 type which extends the Form type.
Here we declare Constructor, Event subs etc.

```
	Type Form1 Extends Form
		Declare Constructor
		
	End Type
```
Next, we have the object's constructors.
Here we have the Form1 constructor which initialize some Form's properties.

```
	Constructor Form1
		' Form1
		With This
			.Name = "Form1"
			.Text = "Form1"
			.Caption = "Form1"
			.SetBounds 0, 0, 350, 300
		End With
	End Constructor
```
After that we define an object var for the type Form1. This var has the name fForm1.

```

Dim Shared fForm1 As Form1

```
After that, we have the app section where we run the app and show the form.

```
	#ifndef _NOT_AUTORUN_FORMS_
		fForm1.Show
		
		App.Run
	#endif
'#End Region
```
Below this section the IDE adds event subs

**Reply by demosthenesk** [reply]
For example:

```
Private Sub Form1.Form_Click_(ByRef Sender As Control)
	*Cast(Form1 Ptr, Sender.Designer).Form_Click(Sender)
End Sub
Private Sub Form1.Form_Click(ByRef Sender As Control)
	Print "Hello World !!!"
End Sub
```
